src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1402,23 +1573,362 @@ DatabaseShardImp::storeLedgerInShard(
+        canAdd_ = false;
+        return boost::none;
+    }
+    if (!sufficientStorage(1, designation, lock))
+    {
+        JLOG(j_.error()) << "insufficient storage space available";
+        canAdd_ = false;
+        return boost::none;
+    }
+
+    return designation;
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath(std::lock_guard<std::mutex> const&) const
+{
+    // If not configured with separate historical paths,
+    // use the main path (dir_) by default.
+    if (historicalPaths_.empty())
+        return dir_;
+
+    boost::filesystem::path historicalShardPath;
+    std::vector<boost::filesystem::path> potentialPaths;
+
+    for (boost::filesystem::path const& path : historicalPaths_)
+    {
+        if (boost::filesystem::space(path).available >= avgShardFileSz_)
+            potentialPaths.push_back(path);
+    }
+
+    if (potentialPaths.empty())
+    {
+        JLOG(j_.error()) << "failed to select a historical shard path";
+        return "";
+    }
+
+    std::sample(
+        potentialPaths.begin(),
+        potentialPaths.end(),
+        &historicalShardPath,
+        1,
+        default_prng());
+
+    return historicalShardPath;
+}
+
+bool
+DatabaseShardImp::checkHistoricalPaths() const
+{
+#if BOOST_OS_LINUX
+    // Each historical shard path must correspond
+    // to a directory on a distinct device or filesystem.
+    // Currently, this constraint is enforced only on
+    // Linux.
+
+    std::unordered_set<int> filesystemIDs;
+    for (auto const& path : historicalPaths_)
+    {
+        struct statvfs buffer;
+        if (statvfs(path.c_str(), &buffer))
+        {
+            JLOG(j_.error())
+                << "failed to acquire stats for a 'historical_shard_path'";
+            return false;
+        }
+
+        if (!filesystemIDs.insert(buffer.f_fsid).second)
+        {
+            JLOG(j_.error())
+                << "each 'historical_shard_path' must correspond to a "
+                   "distinct device or filesystem";
+            return false;
+        }
+    }
+#else
+    // The requirement that each historical storage path
+    // corresponds to a distinct device or filesystem is
+    // enforced only on Linux, so on other platforms
+    // keep track of the available capacities for each
+    // path. Issue a warning if we suspect any of the paths
+    // may violate this requirement.
+
+    // Map byte counts to each path that
+    // shares that byte count.
+    std::unordered_map<std::uint64_t, std::vector<std::string>>
+        uniqueCapacities(historicalPaths_.size());
+
+    for (auto const& path : historicalPaths_)
+    {
+        auto const availableSpace = boost::filesystem::space(path).available;
+
+        if (uniqueCapacities.find(availableSpace) != uniqueCapacities.end())
+            uniqueCapacities[availableSpace].push_back(path.string());
+        else
+            uniqueCapacities[availableSpace] = {path.string()};
```

Comment:
You can do this in one line because the `[]` operator creates a default value (i.e. an empty vector) as soon as you invoke it.
```
uniqueCapacities[boost::filesystem::space(path).available].push_back(path.string());
```

Reply:
Done